### DISCUSSION TITLE
What was the reason not to define == operator for Object class?

### DISCUSSION THREAD

**Discussion by voroninp** [dotnet/csharplang#4150] [Q&A]
I am curios why C# team decided not to define `==` operator for `Object` class which internally would simply call `Object.Equals(left, righ)`.

Why did they prefer to emit `ceq` in this case?

**Comment by CyrusNajmabadi** [comment]
> I am curios why C# team decided not to define == operator for Object class

The C# team does not own System.Object, the BCL does.  However, the reasoning as i understand it is that an overloaded `==` is intended to be provided for types that provide value or value-like semantics.  System.Object does not.

**Reply by voroninp** [reply]
Then why would they allow using `==` which implies this semantics...  I mean - no operator defines in the hierarchy - no valid call.

**Reply by CyrusNajmabadi** [reply]
> Then why would they allow using == which implies this semantics...

`==` does not imply those semantics *by default*.  :)

`==` says: use the semantics provided.  In this case, because there is no overload, it gets the default, which is reference equality, which is what was desired.

> I mean - no operator defines in the hierarchy - no valid call.

No operator defined - default call.
